### ISSUE TITLE
Extremely slow with 7MB schema and 1m LOC on introspection.ts

### ISSUE THREAD

**Issue by deathemperor** [0no-co/gql.tada#22]
### Describe the bug

Also, Typescript errors `Type instantiation is excessively deep and possibly infinite.ts(2589)`

![image](https://github.com/0no-co/gql.tada/assets/4255482/54efb45f-7d16-4b3e-8d42-ecf9f210c622)

### Reproduction

https://github.com/deathemperor/gql.tada/tree/main/examples/example-pokemon-api

### gql.tada version

gql.tada 1.0.2

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/0no-co/gql.tada/discussions) should be used
- [X] Read the [docs](https://gql-tada.0no.co/).
- [X] Follow our [Code of Conduct](https://github.com/0no-co/gql.tada/blob/main/CODE_OF_CONDUCT.md)

**Comment by JoviDeCroock** [comment]
That's a 28.1MB introspection 😅 6/7 times bigger than GitHub, I see that's a Hasura schema so maybe Hasura schema's with their excessive size should be used for our benchmarks rather than the GitHub one.

**Comment by kitten** [comment]
Please note that #25 doesn't actually resolve the “slowness”, but merely makes sure we don't run into TypeScript bailing out on resolving types.

To expand a little bit on this, we’re able to actuall use this schema as an excellent sample to ensure that we can improve the performance of `gql.tada`. However, currently I don’t think we’ll be able to (ever?) make this much faster.

Basically, on my machine, making a change to a query takes 6s to evaluate. Opening the project and letting TSServer process all changes also takes about 6s. We think maybe there’s something going on with re-processing, since a result derived from the introspection may not be properly cached.
That said, even if caching was working, we’d always have this kind of startup time and this noticeably also affects TypeScript in other ways, i.e. even when this types isn't used.

So, just to set expectations here, 28MB for any evaluation is hard, and with the TS type checker there's some constraints which we may not be able to immediately work around. To put this in perspective, if I was to basically print out the minified introspection schema in JSON, I'd get 28,000 pages. It's just really long, even compared to something like the GitHub schema (or other schemas we know about), since Hasura generates a lot of redundant and potentially-unused types and fields in the schema.

So, just to give an update here, the patch, once released (a pre-release is published by our GitHub action) will cause types to be derived for your schema, but it certainly won't be fast for now (and maybe not ever unfortunately 😅)

**Comment by kitten** [comment]
I have to apologise. I was wrong here 😅 

We traced the TypeScript type checker to figure out where it was spending its time.
This took a while because TypeScript’s tracing doesn't actually output anything but a basic view of an expression being checked, most of the time.
However, after going down the call stack and adding more traces, we discovered that `IntrospectionQuery` was redundantly being checked:

<img width="518" alt="Screenshot 2024-01-18 at 19 59 34" src="https://github.com/0no-co/gql.tada/assets/2041385/58005bf0-6fd8-4442-9237-1f68e73be20c">

In other words, we spent a lot of time on making sure that your input `IntrospectionQuery` (i.e. your input schema) matches the `IntrospectionQuery` type. We already knew that this was true though and we rely on this being generated anyway. This was redundantly checked whenever a change was introduced in any file.

Fixing this and removing the redundant checks means that the editing times, even on schemas of this size are now acceptable.

**Comment by kitten** [comment]
[Fixes released in `1.0.3`](https://github.com/0no-co/gql.tada/releases/tag/gql.tada%401.0.3)

**Comment by deathemperor** [comment]
@kitten that was quick! (the responses from you guys lol).

It's much faster now but it still takes about 3s for type suggestions and 3s for re-evaluation. my machine is M3 Max

https://github.com/0no-co/gql.tada/assets/4255482/bbf1df77-d04a-45a4-bf88-71c72c6e007c

**Comment by deathemperor** [comment]
yeah I was waiting for that and just upgraded everything to latest. It works with flying colors now! We're migrating to gql.tada!!

**Comment by deathemperor** [comment]
Fantastic work @kitten I'm so grateful

**Comment by kitten** [comment]
I'm grateful as well ❤️ I really thought for a moment this couldn't be fast, but it was an opportunity for us to dig deeper and to get some insights into why it's slow. So, in the end, we did find that it was fixeable, and I don't think we would've discovered this this soon, if it wasn't for the sample schema 😅 🙌 

On a side-note, I would point out, if you don't want your schema to remain public, I'd set the reproduction repo to be private, so it's out of view again!

**Comment by deathemperor** [comment]
@kitten appreciate your concern, I did intent to delete the repo after the issue is resolved and deleted I have.

Really want to shout out again I understand our usage is unique in many ways (using Hasura isn't new but sharing the schema is :P) and it would probably help. I'm very glad it did.
